File: benchmarks/python/bench_recall.py
#!/usr/bin/env python3
"""Recall latency benchmark — Python binding.

Builds a Store with N items, runs 1000 recall queries, reports p50/p99/p999.
"""

from __future__ import annotations

import statistics
import time
from typing import Any

from smritidb import Store, encode_string

D = 10_000
RECALL_NS = [100, 1_000, 10_000]
RECALL_QUERIES = 1_000
WARMUP_ITERS = 3


def _percentile(samples: list[float], p: float) -> float:
    if not samples:
        return 0.0
    s = sorted(samples)
    idx = round((len(s) - 1) * p)
    return s[min(idx, len(s) - 1)]


def run() -> list[dict[str, Any]]:
    print("[2/4] recall (Python)")
    out: list[dict[str, Any]] = []
    for n in RECALL_NS:
        store = Store(dimension=D)
        t0 = time.perf_counter()
        for i in range(n):
            store.put(f"item_{i}", f"value for item {i}")
        insert_seconds = time.perf_counter() - t0

        # Pre-encode the cues so the timed region measures recall, not encode.
        cues = [encode_string(f"item_{i % n}", D) for i in range(RECALL_QUERIES)]

        # Warm-up.
        for _ in range(WARMUP_ITERS):
            store.recall(cues[0], top_k=10, min_similarity=0.5)

        samples_ms = []
        for c in cues:
            t = time.perf_counter()
            store.recall(c, top_k=10, min_similarity=0.5)
            samples_ms.append((time.perf_counter() - t) * 1000.0)

        p50 = _percentile(samples_ms, 0.50)
        p99 = _percentile(samples_ms, 0.99)
        p999 = _percentile(samples_ms, 0.999)
        print(
            f"  N={n:>5} insert={insert_seconds:.3f}s "
            f"p50={p50:.3f}ms p99={p99:.3f}ms p999={p999:.3f}ms"
        )
        out.append({
            "n": n,
            "queries": len(cues),
            "p50_ms": p50,
            "p99_ms": p99,
            "p999_ms": p999,
            "insert_seconds": insert_seconds,
            "mean_ms": statistics.fmean(samples_ms),
        })
    return out


if __name__ == "__main__":
    import json, sys
    json.dump({"recall": run()}, sys.stdout, indent=2)
    print()

File: benchmarks/typescript/bench-recall.ts
// Recall latency benchmark — TypeScript binding.

import { performance } from "node:perf_hooks";
import { Smritidb, encodeString } from "@tanvrit/smritidb";

import {
  D,
  RECALL_NS,
  RECALL_QUERIES,
  WARMUP_ITERS,
  percentile,
} from "./bench-utils.js";

export interface RecallResult {
  n: number;
  queries: number;
  p50_ms: number;
  p99_ms: number;
  p999_ms: number;
  insert_seconds: number;
}

export function runRecall(): RecallResult[] {
  console.log("[2/4] recall (TypeScript)");
  const out: RecallResult[] = [];

  for (const n of RECALL_NS) {
    const store = new Smritidb({ dimension: D });
    const t0 = performance.now();
    for (let i = 0; i < n; i++) {
      store.put(`item_${i}`, `value for item ${i}`);
    }
    const insertSeconds = (performance.now() - t0) / 1000;

    // Pre-encode cues so the timed region measures recall in isolation.
    const cues: Uint8Array[] = new Array(RECALL_QUERIES);
    for (let i = 0; i < RECALL_QUERIES; i++) {
      cues[i] = encodeString(`item_${i % n}`, D);
    }

    for (let i = 0; i < WARMUP_ITERS; i++) {
      store.recall(cues[0], { topK: 10, minSimilarity: 0.5 });
    }

    const samples_ms: number[] = new Array(cues.length);
    for (let i = 0; i < cues.length; i++) {
      const t = performance.now();
      store.recall(cues[i], { topK: 10, minSimilarity: 0.5 });
      samples_ms[i] = performance.now() - t;
    }

    const p50 = percentile(samples_ms, 0.5);
    const p99 = percentile(samples_ms, 0.99);
    const p999 = percentile(samples_ms, 0.999);
    console.log(
      `  N=${String(n).padStart(5)} insert=${insertSeconds.toFixed(3)}s ` +
        `p50=${p50.toFixed(3)}ms p99=${p99.toFixed(3)}ms p999=${p999.toFixed(3)}ms`,
    );
    out.push({
      n,
      queries: cues.length,
      p50_ms: p50,
      p99_ms: p99,
      p999_ms: p999,
      insert_seconds: insertSeconds,
    });
  }
  return out;
}

if (import.meta.url === `file://${process.argv[1]}`) {
  const out = runRecall();
  console.log(JSON.stringify(out, null, 2));
}
